atomic_multiswap/src/lib.rs

```diff
@@ -0,0 +1,59 @@
+#![no_std]
+
+use soroban_sdk::{contractimpl, contracttype, Account, BytesN, Env, Vec};
+
+mod atomic_swap {
+    soroban_sdk::contractimport!(
+        file = "../target/wasm32-unknown-unknown/release/soroban_atomic_swap_contract.wasm"
+    );
+}
+
+#[derive(Clone)]
+#[contracttype]
+pub struct AccountSwap {
+    pub account: Account,
+    pub amount: i128,
+    pub min_recv: i128,
+}
+
+pub struct AtomicMultiSwapContract;
+
+#[contractimpl]
+impl AtomicMultiSwapContract {
+    // Swap token A for token B atomically. Settle for the minimum requested price
+    // for each party (this is an arbitrary choice to demonstrate the power of
+    // approve; full amounts could be swapped as well).
+    pub fn swap(
+        env: Env,
+        swap_contract: BytesN<32>,
+        token_a: BytesN<32>,
+        token_b: BytesN<32>,
+        accounts_a: Vec<AccountSwap>,
```

Comment:
This contract doesn't call authorize. Does that mean I could front run it and call `swap_client.swap` on a specific pair with the same `Account` before this invocation runs?

Reply:
Yes, that's the tradeoff of using such contracts. There is no negative effect for the user (as the swap would happen on their terms no matter what). The contract should probably `try_swap` to gracefully handle frontrun cases. As it just tries to clear multiple swaps at once, there is really no significantly negative impact if some of them were already cleared (besides some wasted gas).

Reply:
So the thing I'm trying to think about here (and correct me if I'm wrong about anything please) is that after the initial frontrunning invocation, the multiswap invocation will fail because the nonce for one of the accounts has been consumed. You could then submit the individual swap calls because they're signed, but any additional logic in the parent contract wouldn't run. Is this a possible footgun?

Reply:
Ah I misread what you said re: `try_swap`, which would work in this case, but that's a subtle detail I can see contract writers miss.

Reply:
Actually, I'm not sure about this. You might not want to skip a failure due to an insufficient balance, and using `try_swap` does that.

Reply:
Why not? The example here just batches the conceptually independent swaps; arguably in case of failure it probably could even return the accounts we were trying to swap here back to the respective account queues in order to increase the clear probability (e.g. if one of the accounts didn't have the right signature or nonce).

Reply:
> but that's a subtle detail I can see contract writers miss.

That's fair, but I think that it's a general thing to be aware of when batching operations. In any scenario there is a number of potential failure scenarios that may or may not have something to do with auth and hence whether or not handle the failure gracefully is always up to the contract writer.

Reply:
> Why not? The example here just batches the conceptually independent swaps; arguably in case of failure it probably could even return the accounts we were trying to swap here back to the respective account queues in order to increase the clear probability (e.g. if one of the accounts didn't have the right signature or nonce).

In this example, the swaps are independent, but what if I wanted them all to be atomic? I guess the answer is to add authorization to this contract as well.

Reply:
I suppose it depends on your intent. It seems like a better solution for a DEX contract would be to give the approval to the top-level 'router' contract and then delegate implementation. This is more of a toy example of batching (there is really no good reason as to why this should happen on-chain in the first place). But I think it's important that this is still possible to do.
There are also other potential things to do. E.g. imagine that you still ask the users to sign swap, but that's your implementation and it allows itself to only be invoked by a few contracts that you control. This would prevent 'malicious' frontrunning, while still making the user to only sign the swap.